File: OneSignalSDK/onesignal/src/main/java/com/onesignal/HMSLocationController.java
/**
 * Modified MIT License
 * <p>
 * Copyright 2020 OneSignal
 * <p>
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * <p>
 * 1. The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * <p>
 * 2. All copies of substantial portions of the Software may only be used in connection
 * with services provided by OneSignal.
 * <p>
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

package com.onesignal;

import android.location.Location;

import com.huawei.hms.location.FusedLocationProviderClient;
import com.huawei.hms.location.LocationCallback;
import com.huawei.hms.location.LocationRequest;
import com.huawei.hms.location.LocationResult;

class HMSLocationController extends LocationController {

    private static FusedLocationProviderClient hmsFusedLocationClient;
    static LocationUpdateListener locationUpdateListener;

    // Started from this class or PermissionActivity
    static void startGetLocation() {
        initHuaweiLocation();
    }

    private static void initHuaweiLocation() {
        synchronized (syncLock) {
            if (hmsFusedLocationClient == null) {
                try {
                    hmsFusedLocationClient = com.huawei.hms.location.LocationServices.getFusedLocationProviderClient(classContext);
                } catch (Exception e) {
                    OneSignal.Log(OneSignal.LOG_LEVEL.ERROR, "Huawei LocationServices getFusedLocationProviderClient failed! " + e);
                    fireFailedComplete();
                    return;
                }
            }
            if (lastLocation != null)
                fireCompleteForLocation(lastLocation);
            else
                hmsFusedLocationClient.getLastLocation()
                        .addOnSuccessListener(new com.huawei.hmf.tasks.OnSuccessListener<Location>() {
                            @Override
                            public void onSuccess(Location location) {
                                OneSignal.Log(OneSignal.LOG_LEVEL.WARN, "Huawei LocationServices getLastLocation returned location: " + location);
                                if (location == null) {
                                    fireFailedComplete();
                                    return;
                                }
                                lastLocation = location;
                                fireCompleteForLocation(lastLocation);
                                locationUpdateListener = new LocationUpdateListener(hmsFusedLocationClient);
                            }
                        })
                        .addOnFailureListener(new com.huawei.hmf.tasks.OnFailureListener() {
                            @Override
                            public void onFailure(Exception e) {
                                OneSignal.Log(OneSignal.LOG_LEVEL.ERROR, "Huawei LocationServices getLastLocation failed!", e);
                                fireFailedComplete();
                            }
                        });
        }
    }

    static void fireFailedComplete() {
        synchronized (syncLock) {
            hmsFusedLocationClient = null;
        }
    }

    static void onFocusChange() {
        synchronized (syncLock) {
            OneSignal.Log(OneSignal.LOG_LEVEL.DEBUG, "HMSLocationController onFocusChange!");

            // Huawei location not initialized yet
            if (isHMSAvailable() && hmsFusedLocationClient == null)
                return;

            if (hmsFusedLocationClient != null) {
                if (locationUpdateListener != null)
                    hmsFusedLocationClient.removeLocationUpdates(locationUpdateListener);
                locationUpdateListener = new LocationUpdateListener(hmsFusedLocationClient);
            }
        }
    }

    static class LocationUpdateListener extends LocationCallback {

        private FusedLocationProviderClient huaweiFusedLocationProviderClient;

        LocationUpdateListener(FusedLocationProviderClient huaweiFusedLocationProviderClient) {
            this.huaweiFusedLocationProviderClient = huaweiFusedLocationProviderClient;
            init();
        }

        private void init() {
            long updateInterval = BACKGROUND_UPDATE_TIME_MS;
            if (OneSignal.isInForeground())
                updateInterval = FOREGROUND_UPDATE_TIME_MS;

            LocationRequest locationRequest = LocationRequest.create()
                    .setFastestInterval(updateInterval)
                    .setInterval(updateInterval)
                    .setMaxWaitTime((long) (updateInterval * 1.5))
                    .setPriority(LocationRequest.PRIORITY_BALANCED_POWER_ACCURACY);

            OneSignal.Log(OneSignal.LOG_LEVEL.DEBUG, "HMSLocationController Huawei LocationServices requestLocationUpdates!");
            huaweiFusedLocationProviderClient.requestLocationUpdates(locationRequest, this, getLocationHandlerThread().getLooper());
        }

        @Override
        public void onLocationResult(LocationResult locationResult) {
            OneSignal.Log(OneSignal.LOG_LEVEL.DEBUG, "HMSLocationController onLocationResult: " + locationResult);
            if (locationResult != null)
                lastLocation = locationResult.getLastLocation();
        }
    }

}

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/location/internal/controller/impl/HmsLocationController.kt
package com.onesignal.location.internal.controller.impl

import android.location.Location
import android.os.Handler
import android.os.HandlerThread
import com.huawei.hmf.tasks.OnSuccessListener
import com.huawei.hms.location.FusedLocationProviderClient
import com.huawei.hms.location.LocationCallback
import com.huawei.hms.location.LocationRequest
import com.huawei.hms.location.LocationResult
import com.onesignal.core.LogLevel
import com.onesignal.core.internal.application.IApplicationLifecycleHandler
import com.onesignal.core.internal.application.IApplicationService
import com.onesignal.core.internal.common.events.EventProducer
import com.onesignal.core.internal.logging.Logging
import com.onesignal.location.internal.common.LocationConstants
import com.onesignal.location.internal.controller.ILocationController
import com.onesignal.location.internal.controller.ILocationUpdatedHandler
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import java.io.Closeable

// TODO: Implement
class HmsLocationController(
    private val _applicationService: IApplicationService
) : ILocationController {
    private val _locationHandlerThread = LocationHandlerThread()
    private val _startStopMutex = Mutex()
    private val _event = EventProducer<ILocationUpdatedHandler>()

    // exist after start and before stop. updates are protected by the startStopMutex
    private var hmsFusedLocationClient: FusedLocationProviderClient? = null
    private var locationUpdateListener: LocationUpdateListener? = null

    // contains the last location received from location services
    private var _lastLocation: Location? = null

    override suspend fun start(): Boolean {
        var self = this
        var wasSuccessful = false

        withContext(Dispatchers.IO) {
            _startStopMutex.withLock {
                if (hmsFusedLocationClient == null) {
                    try {
                        hmsFusedLocationClient =
                            com.huawei.hms.location.LocationServices.getFusedLocationProviderClient(_applicationService.appContext)
                    } catch (e: Exception) {
                        Logging.error("Huawei LocationServices getFusedLocationProviderClient failed! $e")
                        wasSuccessful = false
                        return@withLock
                    }
                }

                if (_lastLocation != null)
                    _event.fire { it.onLocationChanged(_lastLocation!!) }
                else {
                    var channel = Channel<Boolean>()
                    hmsFusedLocationClient!!.lastLocation
                        .addOnSuccessListener(
                            OnSuccessListener { location ->
                                Logging.warn("Huawei LocationServices getLastLocation returned location: $location")

                                if (location == null) {
                                    runBlocking { channel.send(false) }
                                    return@OnSuccessListener
                                }

                                _lastLocation = location
                                runBlocking { channel.send(true) }
                            }
                        )
                        .addOnFailureListener { e ->
                            Logging.error("Huawei LocationServices getLastLocation failed!", e)
                            runBlocking { channel.send(false) }
                        }
                    wasSuccessful = channel.receive()

                    if (wasSuccessful) {
                        _event.fire { it.onLocationChanged(_lastLocation!!) }
                        locationUpdateListener = LocationUpdateListener(
                            self,
                            _applicationService,
                            hmsFusedLocationClient!!
                        )
                    }
                }
            }
        }

        return wasSuccessful
    }

    override suspend fun stop() {
        _startStopMutex.withLock {
            if (locationUpdateListener != null) {
                locationUpdateListener!!.close()
                locationUpdateListener = null
            }

            if (hmsFusedLocationClient != null) {
                hmsFusedLocationClient = null
            }

            _lastLocation = null
        }
    }

    override fun getLastLocation(): Location? {
        val locationClient: FusedLocationProviderClient = hmsFusedLocationClient ?: return null

        var retVal: Location? = null

        runBlocking {
            var channel = Channel<Any?>()
            locationClient.lastLocation
                .addOnSuccessListener(
                    OnSuccessListener { location ->
                        Logging.warn("Huawei LocationServices getLastLocation returned location: $location")

                        if (location == null) {
                            runBlocking { channel.send(null) }
                            return@OnSuccessListener
                        }

                        retVal = location
                        runBlocking { channel.send(null) }
                    }
                )
                .addOnFailureListener { e ->
                    Logging.error("Huawei LocationServices getLastLocation failed!", e)
                    runBlocking { channel.send(null) }
                }
            channel.receive()
        }

        return retVal
    }

    override fun subscribe(handler: ILocationUpdatedHandler) = _event.subscribe(handler)
    override fun unsubscribe(handler: ILocationUpdatedHandler) = _event.unsubscribe(handler)

    internal class LocationUpdateListener(
        private val _parent: HmsLocationController,
        private val _applicationService: IApplicationService,
        private val huaweiFusedLocationProviderClient: FusedLocationProviderClient
    ) : LocationCallback(), IApplicationLifecycleHandler, Closeable {

        private var hasExistingRequest = false

        init {
            _applicationService.addApplicationLifecycleHandler(this)
            refreshRequest()
        }

        override fun onFocus() {
            Logging.log(LogLevel.DEBUG, "LocationUpdateListener.onFocus()")
            refreshRequest()
        }

        override fun onUnfocused() {
            Logging.log(LogLevel.DEBUG, "LocationUpdateListener.onUnfocused()")
            refreshRequest()
        }

        override fun close() {
            _applicationService.removeApplicationLifecycleHandler(this)

            if (hasExistingRequest)
                huaweiFusedLocationProviderClient.removeLocationUpdates(this)
        }

        override fun onLocationResult(locationResult: LocationResult) {
            Logging.debug("HMSLocationController onLocationResult: $locationResult")
            _parent._lastLocation = locationResult.lastLocation
        }

        private fun refreshRequest() {
            if (hasExistingRequest) {
                huaweiFusedLocationProviderClient.removeLocationUpdates(this)
            }

            var updateInterval = LocationConstants.BACKGROUND_UPDATE_TIME_MS
            if (_applicationService.isInForeground) updateInterval =
                LocationConstants.FOREGROUND_UPDATE_TIME_MS
            val locationRequest = LocationRequest.create()
                .setFastestInterval(updateInterval)
                .setInterval(updateInterval)
                .setMaxWaitTime((updateInterval * 1.5).toLong())
                .setPriority(LocationRequest.PRIORITY_BALANCED_POWER_ACCURACY)
            Logging.debug("HMSLocationController Huawei LocationServices requestLocationUpdates!")
            huaweiFusedLocationProviderClient.requestLocationUpdates(
                locationRequest,
                this,
                _parent._locationHandlerThread.looper
            )
            hasExistingRequest = true
        }
    }

    class LocationHandlerThread internal constructor() :
        HandlerThread("OSH_LocationHandlerThread") {
        var mHandler: Handler

        init {
            start()
            mHandler = Handler(looper)
        }
    }
}
